# <span style="color:red; font-size:75px;"> Báo cáo (Report)</span>

## Nhóm 13
## Thành viên 

1. Nguyễn Minh Sơn - 20110713
2. Lê Anh Nhân - 20110689
3. Đỗ Minh Dũng - 20110620
4. Nguyễn Thái Ngọc Tân - 20110719

# 1. Tóm tắt

Đại dịch covid-19 xuất hiện từ cuối năm 2019, cho đến nay nó đã lan rộng ra hầu hết các quốc gia và vùng lãnh thổ trên thế giới, gây thiệt hại nghiêm trọng về mọi mặt về kinh tế, xã hội....

Mặc dù đã có vaccine phòng ngừa covid-19 nhưng dịch bệnh này vẫn còn hoành hoành ở nhiều quốc gia, liên tục xuất hiện các biến thể mới nguy hiểm.

Vì thế, nhóm đã chọn tập dữ liệu liên quan đến vấn đề này để phân tích tình hình dịch bệnh covid-19. 

Nhóm đã đặt ra một số câu hỏi nghiên cứu về dịch Covid-19, đánh giá xem đại dịch này có phụ thuộc vào các yếu tố như nền kinh tế, dân số hoặc vị trí địa lý của các quốc gia hay không? Tình hình dịch bệnh như thế nào khi biến thể Omicron xuất hiện? Hơn nữa, nhóm cũng muốn đánh giá xem liệu việc tiêm vaccine có làm giảm bớt số ca mắc cũng như tử vong của Covid-19 hay không ?

Để có thể trả lời cho các câu hỏi trên, nhóm đã sử dụng các phương pháp thống kê như kiểm định giả thuyết, các suy diễn về mỗi quan hệ giữa hai biến (C --> Q, Q --> Q ...)

Về kết quả, nhóm đã có thể trả lời được các câu hỏi nghiên cứu đặt ra, tuy nhiên có vài câu nhóm vẫn chưa cảm thấy đúng đắn so với thực tế.

# 2. Giới thiệu

### Câu hỏi nghiên cứu
1. Số ca mắc Covid-19 có phụ thuộc vào các châu lục khi bùng phát hay không ? **(X: Continent, TotalCases)**  **(Y: True/False)**

    ==> Nhóm dựa vào input đầu vào là các biến Continent (biến phân loại), TotalCases (biến số) để xác định xem tổng số ca mắc có phụ thuộc vào các châu lục hay không bằng suy luận thống kê giữa hai biến phân loại và biến số, thực hiện ANOVA F-test (C --> Q)
   
2. Có phải các nước có nền kinh tế phát triển, thu nhập cao thì dịch bệnh Covid-19 sẽ ít nghiêm trọng hơn so với có nền kinh tế phát triển kém hơn, thu nhập thấp hơn không ? **(X: Continent, Location, TotalCases, TotalDeaths)**  **(Y: True/False)**

3. Dân số có phải là nguyên nhân dẫn tới việc gia tăng số ca mắc ở các nước châu Á ? **(X: Population, TotalCases)**  **(Y:TotalCases)**
   
    ==> Nhóm dựa vào input đầu vào là các biến Population (biến số), TotalCases (biến số) để xác định xem có sự phụ thuộc nào giữa dân số và số ca mắc ở các nước châu Á hay không bằng suy luận thống kê giữa hai biến phân loại và biến số (C --> Q)

4. Khi vaccine được phổ biến, điều đó có giúp ích cho việc chống đại dịch Covid-19 trên thế giới? **(X: Date, PeopleFullyVaccinated, NewCases, NewDeaths)**  **(Y: True/False)**

5. Ở các châu lục, trước và sau khi tiêm vaccine, tỉ lệ tử vong của dịch bệnh Covid-19 như thế nào? **(X: Continent, Date, TotalDeaths, TotalCases)**  **(Y: Tỉ lệ tử vong)**

6. Biến thể Omicron xuất hiện vào 24/11/2021 có gây nguy hiểm không ? So sánh với biến thể Delta xuất hiện trước đó ? **(X: Date, NewCases, NewDeaths)**  **(Y: Số ca mắc và tử vong của hai loại biến thể)**

7. So với các bệnh như SARS, EBOLA, MERS, H1N1 thì dịch Covid-19 có nghiêm trọng hơn ? **(X: TotalDeaths, TotalCases)**  **(Y: True/False)**

# 3. Dữ liệu

## 3.1 Mô tả dữ liệu


Link github chứa thư mục data: https://github.com/NMS1010/P4DA_Final_Project

File dữ liệu: owid-covid-data.csv

Giải thích các biến sử dụng

|Biến                                   |Mô tả
|:--------------------------------------|:--------------------------------------------------------------------------|
|`continent`                            |Tên châu lục                                                               |
|`location`                             |Tên nước                                                                   |
|`date`                                 |Ngày quan sát                                                              |
|`total_cases`                          |Tổng số ca mắc covid 19 đã được xác nhận                                   |
|`new_cases`                            |Số ca mắc mới covid 19 theo ngày đã được xác nhận                          |
|`total_cases_per_million`              |Tổng số ca mắc covid 19 đã được xác nhận trên 1,000,000 người              |
|`new_cases_per_million`                |Số ca mắc mới covid 19 theo ngày đã được xác nhận trên 1,000,000 người     |
|`total_deaths`                         |Tổng số ca covid 19 tử vong đã được xác nhận                               |
|`new_deaths`                           |Số ca covid 19 tử vong mới theo ngày đã được xác nhận                      |
|`total_deaths_per_million`             |Tổng số ca covid 19 tử vong đã được xác nhận trên 1,000,000 người          |
|`new_deaths_per_million`               |Số ca covid 19 tử vong mới theo ngày đã được xác nhận trên 1,000,000 người |
|`total_vaccinations`                   |Tổng số liều vaccine                                                       |
|`people_vaccinated`                    |Tổng số người tiêm ít nhất một mũi vaccine                                 |
|`people_fully_vaccinated`              |Tổng số người tiêm đủ liều vaccine theo quy định                           |
|`people_fully_vaccinated_per_hundred`  |Tổng số người tiêm đầy đủ vaccine theo quy định mỗi 100 người              |
|`population`                           |Tổng dân số                                                                |

## 3.2 Tiền xử lý dữ liệu

#### Import library

In [84]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import seaborn as sns
import scipy.stats as stats
from bioinfokit.analys import stat
import statsmodels.api as sm
from statsmodels.formula.api import ols

#### Đọc và làm sạch dữ liệu

In [85]:
covid_df = pd.read_csv('./data/owid-covid-data.csv')

##### Lấy các biến cần thiết

In [86]:
covid_df = covid_df[['continent', 'location', 'date', 'total_cases',  'new_cases', 'total_cases_per_million', 'new_cases_per_million', 'total_deaths', 'new_deaths', 'total_deaths_per_million', 'new_deaths_per_million', 'total_vaccinations', 'people_vaccinated', 'people_fully_vaccinated', 'people_fully_vaccinated_per_hundred', 'population']]

##### Thay đổi tên các biến

In [87]:
covid_df.columns = ['Continent', 'Location', 'Date', 'TotalCases',  'NewCases', 'TotalCases_per_million', 'NewCases_per_million', 'TotalDeaths', 'NewDeaths', 'TotalDeaths_per_million', 'NewDeaths_per_million', 'TotalVaccinations', 'PeopleVaccinated', 'PeopleFullyVaccinated', 'PeopleFullyVaccinated_per_hundred', 'Population']

##### Kiểm tra dữ liệu

In [88]:
covid_df.isnull().sum()

Continent                              9956
Location                                  0
Date                                      0
TotalCases                             3033
NewCases                               3193
TotalCases_per_million                 3791
NewCases_per_million                   3951
TotalDeaths                           20875
NewDeaths                             20839
TotalDeaths_per_million               21620
NewDeaths_per_million                 21584
TotalVaccinations                    121132
PeopleVaccinated                     123339
PeopleFullyVaccinated                126085
PeopleFullyVaccinated_per_hundred    126085
Population                             1075
dtype: int64

Ta thấy vẫn còn nhiều ô nhận giá trị NULL, ta sẽ chuẩn hoá giá trị trong các cột này

##### Các khu vực không có tên châu lục

In [89]:
t = covid_df[covid_df['Continent'].isna()]
t.Location.unique()

array(['Africa', 'Asia', 'Europe', 'European Union', 'High income',
       'International', 'Low income', 'Lower middle income',
       'North America', 'Oceania', 'South America', 'Upper middle income',
       'World'], dtype=object)

Các giá trị trong cột `location` của datatframe thường là tên các quốc gia. Nhưng giá trị trong cột này lại tồn tại tên các châu lục hoặc các liên minh kinh tế trên thế giới (đây là tổng hợp chung tình hình dịch bệnh cho các khu vực đó), nên các cột `continent` tương ứng đã nhận giá trị NaN. Vì thế nhóm sẽ chuẩn hoá các giá trị NaN này thành 0

##### Các khu vực không có thông kê dân số

In [90]:
t = covid_df[covid_df['Population'].isna()]
t.Location.unique()

array(['International', 'Northern Cyprus'], dtype=object)

Một số khu vực trên thế giới như `International`, `Northern Cyprus` chưa có thống kê về dân số trong tập dữ liệu, nên nhóm sẽ bỏ qua các dòng dữ liệU 2 khu vực này.

##### Clean Data

Còn lại những cột có dữ liệu bị để trống (NULL) vì vào thời điểm quan sát thì có thể chưa xảy ra ca tử vong, chưa có ca mắc hoặc chưa có vaccine để tiêm phòng... 

`Vì thế, nhóm sẽ làm sạch dữ liệu của các cột nhận giá trị NaN khác bằng cách điền số 0`

In [91]:
covid_df = covid_df.replace('', np.nan).fillna(0)

Dữ liệu sau khi đã làm sạch

In [92]:
covid_df.isnull().sum()

Continent                            0
Location                             0
Date                                 0
TotalCases                           0
NewCases                             0
TotalCases_per_million               0
NewCases_per_million                 0
TotalDeaths                          0
NewDeaths                            0
TotalDeaths_per_million              0
NewDeaths_per_million                0
TotalVaccinations                    0
PeopleVaccinated                     0
PeopleFullyVaccinated                0
PeopleFullyVaccinated_per_hundred    0
Population                           0
dtype: int64

##### Thời gian thu thập dữ liệu về dịch bệnh

In [93]:
covid_world_df = covid_df[covid_df.Location == 'World']
print(f"Dữ liệu về dịch bệnh Covid-19 được thu thập từ ngày {covid_world_df.Date.min()} tới ngày {covid_world_df.Date.max()}")

Dữ liệu về dịch bệnh Covid-19 được thu thập từ ngày 2020-01-22 tới ngày 2022-03-05


## 3.3 Chi tiết về dữ liệu

### Hình thái dữ liệu

In [94]:
covid_df.shape

(166326, 16)

### 10 quan sát đầu tập dữ liệu

In [95]:
covid_df.head(10)

Continent     Location        Date  TotalCases  NewCases  \
0      Asia  Afghanistan  2020-02-24         5.0       5.0   
1      Asia  Afghanistan  2020-02-25         5.0       0.0   
2      Asia  Afghanistan  2020-02-26         5.0       0.0   
3      Asia  Afghanistan  2020-02-27         5.0       0.0   
4      Asia  Afghanistan  2020-02-28         5.0       0.0   
5      Asia  Afghanistan  2020-02-29         5.0       0.0   
6      Asia  Afghanistan  2020-03-01         5.0       0.0   
7      Asia  Afghanistan  2020-03-02         5.0       0.0   
8      Asia  Afghanistan  2020-03-03         5.0       0.0   
9      Asia  Afghanistan  2020-03-04         5.0       0.0   

   TotalCases_per_million  NewCases_per_million  TotalDeaths  NewDeaths  \
0                   0.126                 0.126          0.0        0.0   
1                   0.126                 0.000          0.0        0.0   
2                   0.126                 0.000          0.0        0.0   
3                   0.126                 0.000          0.0        0.0   
4                   0.126                 0.000          0.0        0.0   
5                   0.126                 0.000          0.0        0.0   
6                   0.126                 0.000          0.0        0.0   
7                   0.126                 0.000          0.0        0.0   
8                   0.126                 0.000          0.0        0.0   
9                   0.126                 0.000          0.0        0.0   

   TotalDeaths_per_million  NewDeaths_per_million  TotalVaccinations  \
0                      0.0                    0.0                0.0   
1                      0.0                    0.0                0.0   
2                      0.0                    0.0                0.0   
3                      0.0                    0.0                0.0   
4                      0.0                    0.0                0.0   
5                      0.0                    0.0                0.0   
6                      0.0                    0.0                0.0   
7                      0.0                    0.0                0.0   
8                      0.0                    0.0                0.0   
9                      0.0                    0.0                0.0   

   PeopleVaccinated  PeopleFullyVaccinated  PeopleFullyVaccinated_per_hundred  \
0               0.0                    0.0                                0.0   
1               0.0                    0.0                                0.0   
2               0.0                    0.0                                0.0   
3               0.0                    0.0                                0.0   
4               0.0                    0.0                                0.0   
5               0.0                    0.0                                0.0   
6               0.0                    0.0                                0.0   
7               0.0                    0.0                                0.0   
8               0.0                    0.0                                0.0   
9               0.0                    0.0                                0.0   

   Population  
0  39835428.0  
1  39835428.0  
2  39835428.0  
3  39835428.0  
4  39835428.0  
5  39835428.0  
6  39835428.0  
7  39835428.0  
8  39835428.0  
9  39835428.0

### 10 quan sát cuối tập dữ liệu

In [96]:
covid_df.tail(10)

Continent  Location        Date  TotalCases  NewCases  \
166316    Africa  Zimbabwe  2022-02-24    234967.0     378.0   
166317    Africa  Zimbabwe  2022-02-25    235467.0     500.0   
166318    Africa  Zimbabwe  2022-02-26    235803.0     336.0   
166319    Africa  Zimbabwe  2022-02-27    235803.0       0.0   
166320    Africa  Zimbabwe  2022-02-28    236380.0     577.0   
166321    Africa  Zimbabwe  2022-03-01    236871.0     491.0   
166322    Africa  Zimbabwe  2022-03-02    237503.0     632.0   
166323    Africa  Zimbabwe  2022-03-03    237503.0       0.0   
166324    Africa  Zimbabwe  2022-03-04    238739.0    1236.0   
166325    Africa  Zimbabwe  2022-03-05    239019.0     280.0   

        TotalCases_per_million  NewCases_per_million  TotalDeaths  NewDeaths  \
166316               15568.801                25.046       5390.0        2.0   
166317               15601.930                33.130       5392.0        2.0   
166318               15624.193                22.263       5393.0        1.0   
166319               15624.193                 0.000       5393.0        0.0   
166320               15662.425                38.232       5395.0        2.0   
166321               15694.959                32.533       5395.0        0.0   
166322               15736.835                41.876       5396.0        1.0   
166323               15736.835                 0.000       5396.0        0.0   
166324               15818.731                81.897       5397.0        1.0   
166325               15837.284                18.553       5397.0        0.0   

        TotalDeaths_per_million  NewDeaths_per_million  TotalVaccinations  \
166316                  357.139                  0.133          7858794.0   
166317                  357.271                  0.133          7870399.0   
166318                  357.338                  0.066          7876942.0   
166319                  357.338                  0.000          7882244.0   
166320                  357.470                  0.133          7890951.0   
166321                  357.470                  0.000          7901360.0   
166322                  357.536                  0.066          7910740.0   
166323                  357.536                  0.000          7921113.0   
166324                  357.603                  0.066          7930621.0   
166325                  357.603                  0.000                0.0   

        PeopleVaccinated  PeopleFullyVaccinated  \
166316         4351621.0              3384590.0   
166317         4355217.0              3388940.0   
166318         4357938.0              3390880.0   
166319         4359872.0              3392578.0   
166320         4362150.0              3396655.0   
166321         4365856.0              3399915.0   
166322         4368726.0              3402434.0   
166323         4372925.0              3406482.0   
166324         4374896.0              3408609.0   
166325               0.0                    0.0   

        PeopleFullyVaccinated_per_hundred  Population  
166316                              22.43  15092171.0  
166317                              22.45  15092171.0  
166318                              22.47  15092171.0  
166319                              22.48  15092171.0  
166320                              22.51  15092171.0  
166321                              22.53  15092171.0  
166322                              22.54  15092171.0  
166323                              22.57  15092171.0  
166324                              22.59  15092171.0  
166325                               0.00  15092171.0

# 4. Trực quan hoá dữ liệu

In [97]:
def plot_bar(df, x, y, title):
    figure = px.bar(df, x = x, y = y, color = 'Continent', text = x, orientation='h', height=500,
                    color_discrete_sequence = px.colors.qualitative.Dark2)
    figure.update_layout(title=title, xaxis_title=x, yaxis_title=y, 
                    yaxis_categoryorder = 'total ascending',
                    uniformtext_minsize=8, uniformtext_mode='hide', template='plotly_dark')
    figure.show()

In [98]:
def plot_hbar(df, col, number, title):
    figure = px.bar(df.sort_values(col).tail(number), x = col, y = 'Location', color = 'Continent', text = col, orientation='h', height=700,
                   color_discrete_sequence = px.colors.qualitative.Dark2)
    figure.update_layout(title=title, xaxis_title=col, yaxis_title="Country", 
                    yaxis_categoryorder = 'total ascending',
                    uniformtext_minsize=8, uniformtext_mode='hide', template='plotly_dark')
    figure.show()

In [99]:
def plot_sub_lines(df, x, y1, y2, title):
    fig = make_subplots(rows=1, cols=2)
    fig.add_trace(
        go.Scatter(x=df[x], y = df[y1], mode="lines", name=y1),
        row = 1, 
        col = 1
    )
    fig.add_trace(
        go.Scatter(x=df[x], y = df[y2], mode="lines", name=y2),
        row = 1, 
        col = 2
    )
    fig.update_layout(height=600, width=1200, title_text=title, template='plotly_dark')
    fig.show()

In [100]:
def plot_map(df, col, title):
    fig = px.scatter_geo(df, locations='Location', locationmode="country names", 
                        color=col, size='Size', 
                        range_color= [0, 10**7], 
                        hover_name='Location',
                        projection="natural earth",     
                        title=title, color_continuous_scale="portland")
    fig.update_layout(
        template='plotly_dark',
        )
    fig.show()

In [101]:
def plot_pie(df, values, names, title):
    fig = px.pie(df, values=values, names=names, title=title)
    fig.update_layout(
        template='plotly_dark',
        )
    fig.show()

In [102]:
def plot_scatter(df, _x, _y, title):
    fig = px.scatter(df, x=_x, y=_y, color = _y,
                 size=_y, hover_name= "Location", title=title, trendline="ols", color_continuous_scale=px.colors.sequential.Viridis)
    fig.show()

In [103]:
covid_group_loction = covid_df.copy()
covid_group_loction = covid_group_loction[covid_group_loction.Continent != 0]
covid_group_loction = covid_group_loction.groupby(['Location'])['TotalCases', 'TotalDeaths'].max()
covid_group_loction = covid_group_loction.reset_index()
covid_group_loction['Size'] = covid_group_loction['TotalCases'].pow(0.3)
plot_map(covid_group_loction, "TotalCases", "Số ca mắc ở các nước trên thế giới")

C:\Users\nguye\AppData\Local\Temp\ipykernel_14380\614326190.py:3: FutureWarning:

Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.



##### Tổng số ca mắc và tử vong mới theo từng ngày trên toàn thế giới kể từ đầu đại dịch


In [104]:
plot_sub_lines(covid_world_df, 'Date', 'NewCases', 'NewDeaths', "Biểu đồ thể hiện tổng số ca mắc và tử vong mới theo ngày")

Vào khoảng cuối năm 2021, biến thể Omicron xuất hiện làm gia tăng nhanh số ca nhiễm mỗi ngày nhưng số ca tử vong theo ngày của biến thể này không cao so với biến thể Delta xuất hiện trước đó. Có lẽ do khi biến thể Omicron xuất hiện thì độ bao phủ vaccine trên thế giới cũng đã tăng cao.

In [105]:
covid_world_df_melt = covid_world_df.melt(id_vars="Date", value_vars=['TotalCases', 'TotalDeaths', 'TotalVaccinations'],
                 var_name='Case', value_name='Count')
fig = px.area(covid_world_df_melt, x="Date", y="Count", color='Case',
title='Tổng số vaccine được tiêm so với tổng số ca mắc và tử vong', color_discrete_sequence = ['orange', 'red','green' ])
fig.update_layout(template='plotly_dark',)
fig.show()

Vaccine xuất hiện vào khoảng đầu năm 2021, từ đó độ bao phủ vaccine trên toàn thế giới ngày càng lớn dần, bỏ xa tổng số ca mắc và số ca tử vong. Vaccine có thể xem công cụ hiệu quả nhất hiện tại , tiến độ tiêm phòng như vậy đang cho thấy tín hiệu tích cực trong việc phòng chống và đi tới chấm dứt đại dịch Covid-19

# 5. Mô hình hoá dữ liệu và kết quả thực nghiệm

## Câu 1: Số ca mắc Covid-19 có phụ thuộc vào các châu lục khi bùng phát hay không ? **(X: Continent)**  **(Y: TotalCases)**

<font face="Cambria Math" size=5 color = #1155cc> Quan hệ 𝐶 → 𝑄: nhiều nhóm độc lập </font>

*  <font face="Cambria Math" size=4> Với 𝐶: Biến Continent (Biến catagorical) </font>
* <font face="Cambria Math" size=4>  Với 𝑄: Biến TotalCases (Biến quantitative)</font>

Tổng số ca mắc theo từng khu vực trên thế giới

In [106]:
covid_country = covid_df.copy()
covid_country = covid_country[covid_country.Continent != 0]
covid_country = covid_country.groupby(by='Location').agg({'Location' : 'first','Continent' : 'first', 'TotalCases': 'max'})
covid_continent = covid_country.groupby(by = 'Continent').agg({'Continent' : 'first', 'TotalCases': 'sum'})
covid_continent

Continent   TotalCases
Continent                                
Africa                Africa   11272358.0
Asia                    Asia  120758420.0
Europe                Europe  161250675.0
North America  North America   93398426.0
Oceania              Oceania    3831683.0
South America  South America   54622339.0

In [107]:
plot_pie(covid_continent,"TotalCases", "Continent", "Tổng số ca mắc Covid-19 theo từng khu vực")

Thông qua biểu đồ trực quan hoá, ta cũng có thể nhận thấy số ca mắc Covid-19 giữa các châu lục khác nhau đều có sự khác nhau, không có sự phụ thuộc vào khu vực cụ thể nào khi dịch bệnh bùng phát. Để có thể kết luận, ta sẽ thực hiện phép thống kê giữa hai biến Continent (catagorical) và biến TotalCases (quantitative) để có thể đưa ra nhận xét. 

In [108]:
def get_value_series(col, continent, data):
    df = data.loc[data["Continent"] == continent]
    case_per_country = list([])
    for value in df[col]:
        case_per_country.append(value)
    return pd.Series(case_per_country)

In [133]:

plot_hbar(covid_country, 'TotalCases', 10, 'Top 10 nước có số ca mắc nhiều nhất trên thế giới')

Số ca mắc nhiều nhất là ở USA, nhưng tập trung nhiều ở các nước châu Âu như Pháp, Anh, Nga, Đức ...

Số ca mắc của từng nước trong từng khu vực trên thế giới

In [109]:
continents = covid_country[covid_country["Continent"] != 0]["Continent"].unique()
dict = {}
for continent in continents:
    serie_values = get_value_series("TotalCases", continent, covid_country)
    dict[continent] = serie_values
case_per_continent = pd.DataFrame(dict)
df_melt = pd.melt(case_per_continent.reset_index(), id_vars=['index'], value_vars=continents)
df_melt.columns = ['index', 'Continent', 'TotalCases']
df_melt

index Continent  TotalCases
0        0      Asia    174331.0
1        1      Asia    421226.0
2        2      Asia    788958.0
3        3      Asia    525171.0
4        4      Asia   1946737.0
..     ...       ...         ...
325     50   Oceania         NaN
326     51   Oceania         NaN
327     52   Oceania         NaN
328     53   Oceania         NaN
329     54   Oceania         NaN

[330 rows x 3 columns]

Ta vẽ biểu đồ hộp Boxplot để xem xét sự phân bố số ca mắc giữa các khu vực trên thế giới

In [110]:
fig = px.box(df_melt, x='Continent', y='TotalCases', title= "Phân bố số ca mắc giữa các khu vực")
fig.update(layout_yaxis_range = [0,9500000])
fig.update_layout(
        template='plotly_dark',
        )
fig.show()

Dựa vào biểu đồ ta thấy số ca mắc phân bố không đồng đều giữa các khu vực và giữa các quốc gia trong khu vực đó. Một số nước có số ca mắc cao hơn các nước khác trong cùng một khu vực

Thực hiện thống kê ANOVA F-test

In [111]:
res = stat()
res.anova_stat(df=df_melt, res_var='TotalCases', anova_model='TotalCases ~ C(Continent)')
res.anova_summary

df        sum_sq       mean_sq         F    PR(>F)
C(Continent)    5.0  4.026032e+14  8.052064e+13  1.718495  0.131427
Residual      219.0  1.026132e+16  4.685532e+13       NaN       NaN

<font face="Cambria Math" size=4 face="black body"> Gọi H0: Số ca mắc Covid-19 trung bình giữa các khu vực như nhau</font>

<font face="Cambria Math" size=4 face="black body"> Gọi Ha: Số ca mắc Covid-19 trung bình giữa các khu vực có sự khác nhau</font>

<font face="Cambria Math" size=4 color = #a61c00> Ta có P-value = 0.131427 thông qua phép thống kê ANOVA F-test </font>

<font face="Cambria Math" size=6 face="black body"> Nhận xét </font>

<font face="Cambria Math" size=4 face="black body">Ta có: P-value = 0.131427 > 0.05 → Ta chấp nhận H0, giữa hai biến Continent và TotalCases không có mối quan hệ. Sự bùng phát dịch bệnh Covid-19 có lẽ đều có thể xảy ra mà không phân biệt châu lục nào.</font>

<font face="Cambria Math" size=4 face="black body">Tuy nhiên, theo biểu đồ thì số ca mắc trung bình giữa các khu vực có sự khác nhau, có lẽ vì ta thực hiện phép thống kê ANOVA F-test trên các nhóm dữ liệu không đồng đều (Số nước của mỗi châu lục là khác nhau và số ca mắc giữa các nước cũng có sự chênh lệch), điều đó có thể giảm độ chính xác của phép thống kê.</font>

## Câu 3: Dân số có phải là nguyên nhân dẫn tới việc gia tăng số ca mắc ở các nước châu Á ? **(X: Population)**  **(Y: TotalCases)**

<font face="Cambria Math" size=5 color = #1155cc> Quan hệ 𝑄 → 𝑄 </font>

*  <font face="Cambria Math" size=4> Với Q: Biến Population (Biến quantitative) </font>
* <font face="Cambria Math" size=4>  Với 𝑄: Biến TotalCases (Biến quantitative)</font>

In [112]:
population_country = covid_df.groupby(by = 'Location').agg({'Continent' : 'first', 'Population': 'max'})
population_country = population_country.groupby(by = "Continent").agg({'Continent' : 'first', 'Population': 'sum'})
population_country = population_country[population_country.Continent != 0]
population_country

Continent    Population
Continent                                 
Africa                Africa  1.371693e+09
Asia                    Asia  4.652616e+09
Europe                Europe  7.506718e+08
North America  North America  5.928348e+08
Oceania              Oceania  4.320246e+07
South America  South America  4.339537e+08

In [113]:
plot_bar(population_country, "Population", "Continent", "Dân số theo khu vực trên thế giới")

Từ biểu đồ trên, ta có thể thấy châu Á là châu lục có số dân đông nhất thế giới, vượt xa các châu lục khác. 

Châu Á là nơi dự đoán có thể có số ca mắc Covid-19 nhiều nhất trên thế giới do dân số đông. 

Ta cùng thực hiện phép thông kê để kiểm chứng liệu dân số đông có phải là nguyên nhân dẫn tới gia tăng số ca mắc hay không ?

In [114]:
cases_country = covid_df.groupby(by = 'Location').agg({'Continent' : 'first','Location' : 'first', 'TotalCases': 'max', 'Population': 'max'})
cases_country = cases_country[cases_country["Continent"] != 0]
cases_Asia = cases_country.loc[cases_country['Continent'] == "Asia"]
cases_Asia.head()

Continent     Location  TotalCases   Population
Location                                                   
Afghanistan      Asia  Afghanistan    174331.0   39835428.0
Armenia          Asia      Armenia    421226.0    2968128.0
Azerbaijan       Asia   Azerbaijan    788958.0   10223344.0
Bahrain          Asia      Bahrain    525171.0    1748295.0
Bangladesh       Asia   Bangladesh   1946737.0  166303494.0

In [115]:
plot_scatter(cases_Asia, "Population", "TotalCases", "Biểu đồ thể hiện sự tương quan giữa dân số và số ca mắc ở các nước châu Á")

***Tính hệ số tương quan để xem mối quan hệ giữa hai biến***

In [116]:
R = np.corrcoef(cases_Asia["Population"], cases_Asia["TotalCases"])
R

array([[1.        , 0.63435655],
       [0.63435655, 1.        ]])

Theo biểu đồ, ta thấy có 2 giá trị ngoại lai ứng với 2 nước là Trung Quốc và Ấn Độ.

Trung Quốc là nước đông dân nhất thế giới, đồng thời là nơi khởi nguồn của dịch bệnh nhưng số ca mắc Covid-19 ở nước này lại không cao. 

Còn trường hợp là Ấn Độ, nước có dân số đông chỉ sau Trung Quốc nhưng số ca mắc cũng tăng nhanh.

Ta thử bỏ điểm dữ liệu này để xem sự khác biệt về hệ số tương quan
 

In [117]:
cases_Asia = cases_Asia.loc[cases_Asia['Location'] != "India"]
cases_Asia = cases_Asia.loc[cases_Asia['Location'] != "China"]

In [118]:
plot_scatter(cases_Asia, "Population", "TotalCases", "Biểu đồ thể hiện sự tương quan giữa dân số và số ca mắc ở các nước châu Á")

***Tính lại hệ số tương quan để xem mối quan hệ giữa hai biến***

In [119]:
R = np.corrcoef(cases_Asia["Population"], cases_Asia["TotalCases"])
R

array([[1.        , 0.50531549],
       [0.50531549, 1.        ]])

Ta thấy hệ số tương quan giảm so với lúc trước khi bỏ hai giá trị ngoại lai

**Ta đặt giả thuyết thống kê**

H0: Hai biến không có mối quan hệ tuyến tính

H1: Hai biến có mối quan hệ tuyến tính

In [120]:
slope, intercept, r_value, p_value, std_err = stats.linregress(cases_Asia["Population"], cases_Asia["TotalCases"])
p_value

0.0002490474547897967

In [121]:
print(f"Ta tính được giá trị P-value = {p_value} < 0.05")
print(f"Ta bác bỏ H0, chấp nhận H1")

Ta tính được giá trị P-value = 0.0002490474547897967 < 0.05
Ta bác bỏ H0, chấp nhận H1


Tuy nhiên, vì hệ số tương quan R = 0.5 không quá mạnh cũng không quá yếu, ta cũng chưa thể kết luận ngay dân số có ảnh hưởng đến số ca mắc Covid-19 ở các quốc gia hay không vì còn nhiều yếu tố ảnh hưởng khác như hệ thống y tế, chính sách cách ly của mỗi quốc gia...

## Câu 6: Biến thể Omicron xuất hiện vào 11/2021 có gây nguy hiểm không? So sánh với biến thể delta xuất hiện trước đó

In [128]:
covid_compare = covid_df[['Date', 'NewCases', 'NewDeaths']]
covid_compare['Date'] = pd.to_datetime(covid_compare['Date'])
covid_compare['Month_Years']=pd.to_datetime(covid_compare['Date']).dt.strftime('%Y-%m')
covid_compare

C:\Users\nguye\AppData\Local\Temp\ipykernel_14380\2087622515.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

C:\Users\nguye\AppData\Local\Temp\ipykernel_14380\2087622515.py:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Date  NewCases  NewDeaths Month_Years
0      2020-02-24       5.0        0.0     2020-02
1      2020-02-25       0.0        0.0     2020-02
2      2020-02-26       0.0        0.0     2020-02
3      2020-02-27       0.0        0.0     2020-02
4      2020-02-28       0.0        0.0     2020-02
...           ...       ...        ...         ...
166321 2022-03-01     491.0        0.0     2022-03
166322 2022-03-02     632.0        1.0     2022-03
166323 2022-03-03       0.0        0.0     2022-03
166324 2022-03-04    1236.0        1.0     2022-03
166325 2022-03-05     280.0        0.0     2022-03

[166326 rows x 4 columns]

In [131]:
covid_compare.dtypes

Date           datetime64[ns]
NewCases              float64
NewDeaths             float64
Month_Years            object
dtype: object

## Câu 7: So với các bệnh như SARS, EBOLA, MERS, H1N1 thì dịch Covid-19 có nghiêm trọng hơn ?

In [122]:
epidemics = pd.DataFrame({
    'Dịch bệnh' : ['COVID-19', 'SARS', 'EBOLA', 'MERS', 'H1N1'],
    'Năm bắt đầu' : [2019, 2003, 2014, 2012, 2009],
    'Năm kết thúc' : ["2022 (giả sử)", 2004, 2016, 2017, 2010],
    'Tổng ca mắc' : [covid_world_df['TotalCases'].max(), 8096, 28646, 2494, 6724149],
    'Tổng ca tử vong' : [covid_world_df['TotalDeaths'].max(), 774, 11323, 858, 19654]
})

epidemics['Tỉ lệ tử vong'] = round((epidemics['Tổng ca tử vong']/epidemics['Tổng ca mắc'])*100, 2)
epidemics

Dịch bệnh  Năm bắt đầu   Năm kết thúc  Tổng ca mắc  Tổng ca tử vong  \
0  COVID-19         2019  2022 (giả sử)  445129499.0        5995245.0   
1      SARS         2003           2004       8096.0            774.0   
2     EBOLA         2014           2016      28646.0          11323.0   
3      MERS         2012           2017       2494.0            858.0   
4      H1N1         2009           2010    6724149.0          19654.0   

   Tỉ lệ tử vong  
0           1.35  
1           9.56  
2          39.53  
3          34.40  
4           0.29

In [123]:
epidemics_melt = epidemics.melt(id_vars='Dịch bệnh', value_vars=['Tổng ca mắc', 'Tổng ca tử vong', 'Tỉ lệ tử vong'],
                      var_name='Case', value_name='Value')

fig = px.bar(epidemics_melt, x="Dịch bệnh", y="Value", color='Dịch bệnh', text='Value', facet_col="Case",
             color_discrete_sequence = px.colors.qualitative.Bold)
fig.update_traces(textposition='outside')
fig.update_layout(uniformtext_minsize=8, uniformtext_mode='hide',template='plotly_dark')
fig.update_yaxes(showticklabels=False)
fig.layout.yaxis2.update(matches=None)
fig.layout.yaxis3.update(matches=None)
fig.show()

Từ biểu đồ ta có thể thấy rằng số ca mắc và tử vong của đại dịch Covid-19 là rất lớn, vượt xa so với các đại dịch khác. Tuy nhiên, tỉ lệ tử vong lại không cao, một phần vì vaccine xuất hiện giúp hạn chế đi số ca tử vong này. 

Chúng ta vẫn chưa biết khi nào đại dịch Covid-19 sẽ kết thúc, vì thế hãy giữ an toàn cho chính bản thân mình và mọi người xung quanh